Google.Api.Gax.Grpc/ChannelFactory.cs

```diff
@@ -0,0 +1,21 @@
+﻿/*
+ * Copyright 2019 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+using Grpc.Core;
+
+namespace Google.Api.Gax.Grpc
+{
+    // TODO: Make this an abstract class instead of a delegate?
+
+    /// <summary>
+    /// Creates a channel for the given endpoint, with the given credentials and options.
+    /// </summary>
+    /// <param name="endpoint">The endpoint to connect to. Must not be null.</param>
+    /// <param name="credentials">Credentials for the channel. Must not be null.</param>
+    /// <param name="options">Options for the channel. Must not be null.</param>
+    /// <returns>A channel for the given endpoint.</returns>
+    public delegate ChannelBase ChannelFactory(ServiceEndpoint endpoint, ChannelCredentials credentials, GrpcChannelOptions options);
```

Comment:
I assume both Grpc.Core and grpc-dotnet would provide an implementation of such factory that creates channels of the right kind. If that's so and the channelFactory implementation is the only piece that would depend on Grpc.Core or Grpc.Net.Client, it should at least be trivial to generate two versions of the nuget:
- one version depends on Grpc.Core package
- the other version depends on Grpc.Net.Client
- they are exactly the same except one file that provides the right channelFactory implementation (there could even be an #ifdef that allows one to switch by defining a symbol)

This is still not an ideal soution of course, but at least it would make it reasonably easy to obtain packages that only depend on Grpc.Net.Client

Reply:
I think if we were able to generate a set of client libraries that depended on Grpc.Net.Client instead of Grpc.Core (and would be able to upload them to dev nuget channel), it would be huge for making sure that one day, the managed grpc client will work with google APIs.
(having a package to play with makes it much easier to see what features are missing a what bugs need to be fixed).

Reply:
Actually only 2 pieces which depend on the managed/native part are `ChannelPool` and `ClientBuilderBase<T>` if they both moved to a separate package it would solve all your questions and concerns (in particular: `Client.ShutdownAsync` vs `Client.Dispose`, unsupported client options (keepalive), etc),

Reply:
I doubt that majority of the "end" libraries would need to have a ClientBuilder or a ChannelPool. So, I think it is _a good enough_ solution.

Reply:
The GAX libraries are specifically designed to support the libraries at https://github.com/googleapis/google-cloud-dotnet, and *every* gRPC-based library there uses `ClientBuilder`; almost all of them use `ChannelPool`. So no, it doesn't really solve much.
If you know of other libraries using GAX, I'd be interested to hear about them.

Reply:
More details about the deployment issue. The absolute (and unachievable as far as I'm aware) ideal that I'd like is to have something which is effectively "default to Grpc.Core, but allow an opt out somehow". Basically, there has to be a dependency on Grpc.Core somewhere, but the question is where.

**Option 1: Bias heavily towards Grpc.Core**

Google.Cloud.Speech.V1 depends on Grpc.Core (as it currently does, both directly and indirectly). It tells GAX that it wants to use the Grpc.Core-based way of creating clients and shutting down channels. That leaves two problems if you want to use Grpc.Net.Client instead:

- Grpc.Core (which is large) will *always* get deployed alongside the application, even if you're not using it.
- There has to be some way of telling Google.Cloud.Speech.V1 to use the different abstraction. It's not obvious where that injection point should be.

**Option 2: Three packages per API**

An alternative would be to make Google.Cloud.Speech.V1 only depend on GAX (and indirectly Grpc.Core.Api), and to produce two *extra* packages: Google.Cloud.Speech.V1.GrpcCore and Google.Cloud.Speech.V1.GrpcNetClient. It's still not obvious to me what the API surface looks like for that, and it means tripling the number of packages we produce and maintain, which isn't ideal either.

**Option 3: Two packages per API**

This is basically Jan's suggestion above. It involves code duplication (there being two different SpeechClient classes, for example - one in each flavour of package) but does reduce the dependency chain somewhat.

If the purpose of using Grpc.Net.Client is primarily to validate that we have compatibility, then we should probably go for option 1 listed above - the unnecessary "baggage" of Grpc.Core probably isn't much of an issue for that, although we still need to work out the injection point for the abstraction. (A static property would be feasible but unpleasant, for example.)

I don't have a good answer to all of this yet, but that's a summary of the problem I'm trying to get my head round.

Reply:
I think for now we mostly want to:
- validate that Grpc.Net.Client works for Google API (and see what doesn't work)
- we already know that at least in some cases Grpc.Net.Client wont works because it doesn't support all the necessary features (e.g. some channel options are not implemented).
- there's some potential that Grpc.Net.Client might become the default choice one day, but it's likely a distant future and not clear if it will ever happen at all.

So I'd say we can go with Option 1 for now, assuming we can provide a simple way of using Grpc.Net.Client with the client libraries for those who want to opt-in (for testing, validation or to work on filling in the gaps in Grpc.Net.Client).
In addition to that, we could provide
- a HOWTO doc for using client libraries with Grpc.Net.Client (and list of gotchas)
- a small support nuget package that provides the Grpc.Net.Client binding in a digestible form (to make using it as simple as possible).

WDYT?

CC @JamesNK

Reply:
I'll experiment with it - now is definitely the time to do so. I'm torn between "let this package depend on Grpc.Core to streamline the normal case" and "provide a Google.Api.Gax.Grpc.GrpcCore package that has the Grpc.Core implementation of the abstraction layer". The latter is sort of cleaner, but requires yet another package to maintain. I think I'll need to play a bit more.

Reply:
I'm now thinking of a *possible* option of loading things with reflection, so that we can avoid extra packages. It may make things tricky in terms of package trimming, but it's at least worth experimenting with.